Generated C has typedefs in the wrong order

### Describe the bug

I have an auto translated wrapper that compiles successfully when in the main module, but does not compile correctly when its in its own module.

### Reproduction Steps

Using the attached module, and the following code:
```
module main

import gdi

[export: 'hello_extension_entry']
pub fn hello_extension_entry(gpaddr gdi.GDExtensionInterfaceGetProcAddress, clp gdi.GDExtensionClassLibraryPtr, mut gdinit gdi.GDExtensionInitialization) gdi.GDExtensionBool {
    return 1
}
```

Its meant to be compiled as a shared library: `v -shared .`

### Expected Behavior

Successful compile

### Current Behavior

When compiling, it gives the following error:
```
v -no-parallel -shared .    
==================
c:/users/jcwea/documents/git/v/thirdparty/tcc/include/winapi/synchapi.h:167: warning: CreateMutex redefined
C:/Users/jcwea/AppData/Local/Temp/v_0/jvnet.4895371472969035013.tmp.so.c:2109: error: identifier expected
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.
```
When checking the tmp c file on line 2109, it shows it using `gdi__GDExtensionInterfaceFunctionPtr`, which is defined on line 2172:
line 2109:
![image](https://github.com/vlang/v/assets/6415281/eed62753-5e08-42ae-acc6-717e2bb253c5)
line 2172:
![image](https://github.com/vlang/v/assets/6415281/0230e9fc-3d8b-4c65-8646-d54c53f6cc00)



### Possible Solution

typedefs could be checked for types it uses, and if that type hasn't been defined yet, define it first

### Additional Information/Context

This is for GDExtension 4.1 for Godot. I plan on fully developing and publishing this as an open source library once its in a usable state, allowing users to use V for game development in Godot

### V version

V 0.4.1 bb95a27

### Environment details (OS name and version, etc.)

Tested on Windows:
```
v doctor
V full version: V 0.4.1 d93488b.bb95a27
OS: windows, Microsoft Windows 11 Pro Insider Preview v23541 64-bit
Processor: 24 cpus, 64bit, little endian, 

getwd: C:\Users\jcwea\Documents\git\jvnet
vexe: C:\Users\jcwea\Documents\git\v\v.exe
vexe mtime: 2023-09-16 16:23:44

vroot: OK, value: C:\Users\jcwea\Documents\git\v
VMODULES: OK, value: C:\Users\jcwea\.vmodules
VTMP: OK, value: C:\Users\jcwea\AppData\Local\Temp\v_0

Git version: git version 2.40.0.windows.1
Git vroot status: weekly.2023.37-22-gbb95a276-dirty
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,
operable program or batch file.

thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```

And WSL
```
v doctor
V full version: V 0.4.1 bfaa3de.bb95a27
OS: linux, Linux version 5.15.90.4-microsoft-standard-WSL2 (root@4b885f886b45) (gcc (GCC) 11.2.0, GNU ld (GNU Binutils) 2.37) #1 SMP Tue Jul 18 21:28:32 UTC 2023 (WSL 2)
Processor: 24 cpus, 64bit, little endian, 13th Gen Intel(R) Core(TM) i7-13700K

getwd: /mnt/c/Users/jcwea/Documents/git/jvnet
vexe: /home/jc/git/v/v
vexe mtime: 2023-09-16 16:25:34

vroot: OK, value: /home/jc/git/v
VMODULES: OK, value: /home/jc/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.42.0
Git vroot status: weekly.2023.32-234-gbb95a276
.git/config present: true

CC version: cc (GCC) 13.2.1 20230801
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```
> [!IMPORTANT]
> You can vote for this issue using the 👍 reaction. More votes increase the issue's priority
> for developers.
>
> Take into account that only the 👍 reaction counts as a vote.
> Only reactions to the issue itself will be counted as votes, not comments.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -1515,11 +1515,7 @@ static inline void __${sym.cname}_pushval(${sym.cname} ch, ${push_arg} val) {
 			g.write_alias_typesymbol_declaration(sym)
 		}
 	}
-	for sym in g.table.type_symbols {
-		if sym.kind == .function && sym.name !in c.builtins {
-			g.write_fn_typesymbol_declaration(sym)
-		}
-	}
+	g.write_sorted_fn_typesymbol_declaration()
 	// Generating interfaces after all the common types have been defined
 	// to prevent generating interface struct before definition of field types
 	for sym in g.table.type_symbols {
@@ -6367,6 +6363,63 @@ fn (mut g Gen) sort_globals_consts() {
 			if node.value == order {
 				g.sorted_global_const_names << node.name
 			}
+		}
+	}
+}
+
+// sort functions by dependent arguments and return value
+// As functions may depend on one another, make sure they are
+// defined in the correct order: add non dependent ones first.
+fn (mut g Gen) write_sorted_fn_typesymbol_declaration() {
+	util.timing_start(@METHOD)
+	defer {
+		util.timing_measure(@METHOD)
+	}
+	mut syms := []&ast.TypeSymbol{} // functions to be defined
+	for sym in g.table.type_symbols {
+		if sym.kind == .function && sym.name !in c.builtins {
+			syms << sym
+		}
+	}
+	mut pending := []&ast.TypeSymbol{} // functions with a dependency
+	for {
+		// Add non dependent functions or functions which
+		// dependency has been added.
+		next: for sym in syms {
+			info := sym.info as ast.FnType
+			func := info.func
+			return_sym := g.table.sym(func.return_type)
+			if return_sym in syms {
+				pending << sym
+				continue
+			}
+			for param in func.params {
+				param_sym := g.table.sym(param.typ)
+				if param_sym in syms {
+					pending << sym
+					continue next
+				}
+			}
+			g.write_fn_typesymbol_declaration(sym)
+		}
+		if pending.len == 0 {
+			// All functions were added.
+			break
+		}
+		if syms.len == pending.len {
+			// Could not add any function: there is a circular
+			// dependency.
+			mut deps := []string{}
+			for sym in pending {
+				deps << sym.name
+			}
+			verror(
+				'cgen.write_sorted_fn_typesymbol_declaration(): the following functions form a dependency cycle:\n' +
+				deps.join(','))
+		}
+		unsafe {
+			// Swap the to-be-processed and the dependent functions.
+			syms, pending = pending, syms[..0]
 		}
 	}
 }

--- a/vlib/v/gen/c/testdata/func_type_dependency.c.must_have
+++ b/vlib/v/gen/c/testdata/func_type_dependency.c.must_have
@@ -1,0 +1,4 @@
+typedef void (*gdi__Function_ptr)();
+typedef void (*gdi__Function_ptr2)();
+typedef gdi__Function_ptr* (*gdi__Get_proc_address)(i8*);
+typedef void (*gdi__Set_proc_address)(gdi__Function_ptr2*);

--- a/vlib/v/gen/c/testdata/func_type_dependency.vv
+++ b/vlib/v/gen/c/testdata/func_type_dependency.vv
@@ -1,0 +1,8 @@
+// vtest vflags: -shared
+module gdi
+
+pub type Get_proc_address = fn(&i8) &Function_ptr
+pub type Set_proc_address = fn(&Function_ptr2)
+
+pub type Function_ptr = fn ()
+pub type Function_ptr2 = fn ()

